servant-server/src/Servant/Server/Internal.hs

```diff
@@ -718,8 +720,8 @@ instance ( AllCTUnrender list a, HasServer sublayout
       mrqbody <- handleCTypeH (Proxy :: Proxy list) (cs contentTypeH)
              <$> lazyRequestBody request
       case mrqbody of
-        Nothing -> return $ failWith $ UnsupportedMediaType
-        Just (Left e) -> return $ failWith $ InvalidBody e
+        Nothing -> return $ FailFatal err415
+        Just (Left e) -> return $ FailFatal err400 { errBody = cs e }
```

Comment:
Here there's the added advantage that we can get rid of request body memoization, which to my mind seems like a big win. And this gives a somewhat more principled notion of error messages - there can only be one endpoint that can ever deserialize the body, so you always get the correct error message from deserialization errors.

 On the other hand, there isn't the argument (that I think there is with content-types) about it actually leading to better design.